src/consensus/proposer.rs

```diff
@@ -415,9 +447,58 @@ impl Proposer for BlockProposer {
     }
 
     fn add_proposed_value(&mut self, full_proposal: &FullProposal) -> Validity {
-        if let Some(proto::full_proposal::ProposedValue::Block(_block)) =
+        if let Some(proto::full_proposal::ProposedValue::Block(block)) =
             &full_proposal.proposed_value
         {
+            let header = block.header.as_ref().unwrap();
+            let height = header.height.unwrap();
+
+            if height != self.get_confirmed_height().increment() {
+                error!(
+                    shard = height.shard_index,
+                    our_height = height.block_number,
+                    proposal_height = height.block_number,
+                    "Cannot validate height, not the next height"
+                );
+                return Validity::Invalid;
+            }
+
+            if header.chain_id != (self.chain_id as i32) {
+                error!("Received block with wrong chain_id: {}", header.chain_id);
+                return Validity::Invalid;
+            }
+            if header.version != PROTOCOL_VERSION {
+                error!(
+                    "Received block with wrong protocol version: {}",
+                    header.version
+                );
+                return Validity::Invalid;
+            }
+            if header.height.is_none() {
+                error!("Received block with missing height");
+                return Validity::Invalid;
+            }
+            if header.shard_witnesses_hash.is_empty() {
+                error!("Received block with missing shard witnesses hash");
+                return Validity::Invalid;
+            }
+            if block.shard_witness.is_none() {
+                error!("Received block with missing shard witnesses");
+                return Validity::Invalid;
+            }
+            let witness = block.shard_witness.as_ref().unwrap();
+            if witness.shard_chunk_witnesses.len() != self.num_shards as usize {
```

Comment:
We need unit tests for the proposer now, let me look into it. I might just do it in a separate pr.